In [ ]:
pip install pandas openpyxl

Note: you may need to restart the kernel to use updated packages.


# Setting Up

In [ ]:
import pandas as pd


# Setting to make number easier to read on display
pd.options.display.float_format = '{:20.2f}'.format

In [4]:
# Import dataset
df = pd.read_excel("./online+retail+ii/online_retail_II.xlsx", sheet_name = 0) # only read the first sheet
df.head(10)

Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434    79323P                   PINK CHERRY LIGHTS        12   
2  489434    79323W                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
5  489434     22064           PINK DOUGHNUT TRINKET POT         24   
6  489434     21871                  SAVE THE PLANET MUG        24   
7  489434     21523   FANCY FONT HOME SWEET HOME DOORMAT        10   
8  489435     22350                            CAT BOWL         12   
9  489435     22349       DOG BOWL , CHASING BALL DESIGN        12   

          InvoiceDate                Price          Customer ID  \
0 2009-12-01 07:45:00                 6.95             13085.00   
1 2009-12-01 07:45:00                 6.75             13085.00   
2 2009-12-01 07:45:00                 6.75             13085.00   
3 2009-12-01 07:45:00                 2.10             13085.00   
4 2009-12-01 07:45:00                 1.25             13085.00   
5 2009-12-01 07:45:00                 1.65             13085.00   
6 2009-12-01 07:45:00                 1.25             13085.00   
7 2009-12-01 07:45:00                 5.95             13085.00   
8 2009-12-01 07:46:00                 2.55             13085.00   
9 2009-12-01 07:46:00                 3.75             13085.00   

          Country  
0  United Kingdom  
1  United Kingdom  
2  United Kingdom  
3  United Kingdom  
4  United Kingdom  
5  United Kingdom  
6  United Kingdom  
7  United Kingdom  
8  United Kingdom  
9  United Kingdom

# Exploratory Data Analysis

In [5]:
# Examine high-level view of our data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 525461 entries, 0 to 525460
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      525461 non-null  object        
 1   StockCode    525461 non-null  object        
 2   Description  522533 non-null  object        
 3   Quantity     525461 non-null  int64         
 4   InvoiceDate  525461 non-null  datetime64[ns]
 5   Price        525461 non-null  float64       
 6   Customer ID  417534 non-null  float64       
 7   Country      525461 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 32.1+ MB


In [6]:
# Examine high-level view of numeric data
df.describe()

Quantity                    InvoiceDate  \
count            525461.00                         525461   
mean                 10.34  2010-06-28 11:37:36.845017856   
min               -9600.00            2009-12-01 07:45:00   
25%                   1.00            2010-03-21 12:20:00   
50%                   3.00            2010-07-06 09:51:00   
75%                  10.00            2010-10-15 12:45:00   
max               19152.00            2010-12-09 20:01:00   
std                 107.42                            NaN   

                     Price          Customer ID  
count            525461.00            417534.00  
mean                  4.69             15360.65  
min              -53594.36             12346.00  
25%                   1.25             13983.00  
50%                   2.10             15311.00  
75%                   4.21             16799.00  
max               25111.09             18287.00  
std                 146.13              1680.81

In [7]:
# Examine high-level view of non-numeric data
df.describe(include = 'O')

Invoice StockCode                         Description         Country
count    525461    525461                              522533          525461
unique    28816      4632                                4681              40
top      537434    85123A  WHITE HANGING HEART T-LIGHT HOLDER  United Kingdom
freq        675      3516                                3549          485852

In [8]:
# Examine missing customer ID data
df[df["Customer ID"].isna()].head(10)



Invoice StockCode                   Description  Quantity  \
263   489464     21733                  85123a mixed       -96   
283   489463     71477                         short      -240   
284   489467    85123A                   21733 mixed      -192   
470   489521     21646                           NaN       -50   
577   489525    85226C     BLUE PULL BACK RACING CAR         1   
578   489525     85227   SET/6 3D KIT CARDS FOR KIDS         1   
1055  489548     22271          FELTCRAFT DOLL ROSIE         1   
1056  489548     22254          FELT TOADSTOOL LARGE        12   
1057  489548     22273          FELTCRAFT DOLL MOLLY         3   
1058  489548     22195  LARGE HEART MEASURING SPOONS         1   

             InvoiceDate                Price  Customer ID         Country  
263  2009-12-01 10:52:00                 0.00          NaN  United Kingdom  
283  2009-12-01 10:52:00                 0.00          NaN  United Kingdom  
284  2009-12-01 10:53:00                 0.00          NaN  United Kingdom  
470  2009-12-01 11:44:00                 0.00          NaN  United Kingdom  
577  2009-12-01 11:49:00                 0.55          NaN  United Kingdom  
578  2009-12-01 11:49:00                 0.85          NaN  United Kingdom  
1055 2009-12-01 12:32:00                 2.95          NaN  United Kingdom  
1056 2009-12-01 12:32:00                 1.25          NaN  United Kingdom  
1057 2009-12-01 12:32:00                 2.95          NaN  United Kingdom  
1058 2009-12-01 12:32:00                 1.65          NaN  United Kingdom

##### Even there are records with legit transactions, when we don't have a customer ID, we can't categorize or effectively work on that data because it's near impossible for us to infer which customer made this purchase so the strategy when we get to cleaning is to drop the null customer ID

In [9]:
# Interestingly, there are negative quantities so let's see what that data looks like
df[df['Quantity'] <0 ].head(10)

Invoice StockCode                        Description  Quantity  \
178  C489449     22087           PAPER BUNTING WHITE LACE       -12   
179  C489449    85206A       CREAM FELT EASTER EGG BASKET        -6   
180  C489449     21895      POTTING SHED SOW 'N' GROW SET        -4   
181  C489449     21896                 POTTING SHED TWINE        -6   
182  C489449     22083         PAPER CHAIN KIT RETRO SPOT       -12   
183  C489449     21871                SAVE THE PLANET MUG       -12   
184  C489449     84946    ANTIQUE SILVER TEA GLASS ETCHED       -12   
185  C489449    84970S  HANGING HEART ZINC T-LIGHT HOLDER       -24   
186  C489449     22090          PAPER BUNTING RETRO SPOTS       -12   
196  C489459    90200A         PURPLE SWEETHEART BRACELET        -3   

            InvoiceDate                Price          Customer ID  \
178 2009-12-01 10:33:00                 2.95             16321.00   
179 2009-12-01 10:33:00                 1.65             16321.00   
180 2009-12-01 10:33:00                 4.25             16321.00   
181 2009-12-01 10:33:00                 2.10             16321.00   
182 2009-12-01 10:33:00                 2.95             16321.00   
183 2009-12-01 10:33:00                 1.25             16321.00   
184 2009-12-01 10:33:00                 1.25             16321.00   
185 2009-12-01 10:33:00                 0.85             16321.00   
186 2009-12-01 10:33:00                 2.95             16321.00   
196 2009-12-01 10:44:00                 4.25             17592.00   

            Country  
178       Australia  
179       Australia  
180       Australia  
181       Australia  
182       Australia  
183       Australia  
184       Australia  
185       Australia  
186       Australia  
196  United Kingdom

In [29]:
# It is unclear if all the invoices with negative quantity have "C" in their invoice. Let's validate it
df_negative_quantity = df[df['Quantity'] < 0]
df_negative_quantity.loc[:, "Invoice"] = df_negative_quantity["Invoice"].astype("str")
df_negative_quantity[df_negative_quantity["Invoice"].str.match("^C\\d{6}$") == False]

Invoice StockCode   Description  Quantity         InvoiceDate  \
263     489464     21733  85123a mixed       -96 2009-12-01 10:52:00   
283     489463     71477         short      -240 2009-12-01 10:52:00   
284     489467    85123A   21733 mixed      -192 2009-12-01 10:53:00   
470     489521     21646           NaN       -50 2009-12-01 11:44:00   
3114    489655     20683           NaN       -44 2009-12-01 17:26:00   
...        ...       ...           ...       ...                 ...   
524458  538092     37467           NaN      -177 2010-12-09 14:49:00   
525231  538159     21324           NaN       -18 2010-12-09 17:17:00   
525232  538158     20892           NaN       -32 2010-12-09 17:17:00   
525234  538161    46000S  Dotcom sales      -100 2010-12-09 17:25:00   
525235  538162    46000M  Dotcom sales      -100 2010-12-09 17:25:00   

                      Price  Customer ID         Country  
263                    0.00          NaN  United Kingdom  
283                    0.00          NaN  United Kingdom  
284                    0.00          NaN  United Kingdom  
470                    0.00          NaN  United Kingdom  
3114                   0.00          NaN  United Kingdom  
...                     ...          ...             ...  
524458                 0.00          NaN  United Kingdom  
525231                 0.00          NaN  United Kingdom  
525232                 0.00          NaN  United Kingdom  
525234                 0.00          NaN  United Kingdom  
525235                 0.00          NaN  United Kingdom  

[2121 rows x 8 columns]

In [30]:
df_negative_quantity_notnullID = df[(df['Quantity'] < 0) & (df['Customer ID'].notnull())]
df_negative_quantity_notnullID.loc[:, "Invoice"] = df_negative_quantity_notnullID["Invoice"].astype("str")
df_negative_quantity_notnullID[df_negative_quantity_notnullID["Invoice"].str.match("^C\\d{6}$") == False]

Empty DataFrame
Columns: [Invoice, StockCode, Description, Quantity, InvoiceDate, Price, Customer ID, Country]
Index: []

##### Invoices with negative quantity can either have C in their invoice with non-null customer ID, or null customer ID with no C in their invoice

In [31]:
# Examine if C is the only unique letter that appears in front of an invoice number
df["Invoice"].str.replace("[0-9]", "", regex = True).unique()

array([nan, 'C', 'A'], dtype=object)

In [32]:
# There is a letter "A" beside "C" in front of an invoice number. Let's see what these are
df[df["Invoice"].str.startswith("A", na= False)]

Invoice StockCode      Description  Quantity         InvoiceDate  \
179403  A506401         B  Adjust bad debt         1 2010-04-29 13:36:00   
276274  A516228         B  Adjust bad debt         1 2010-07-19 11:24:00   
403472  A528059         B  Adjust bad debt         1 2010-10-20 12:04:00   

                      Price  Customer ID         Country  
179403            -53594.36          NaN  United Kingdom  
276274            -44031.79          NaN  United Kingdom  
403472            -38925.87          NaN  United Kingdom

In [33]:
# StockCode is unusual as it should comprise 5-digit number (stated in the documentation), instead of letter "B". Examine the column StockCode
df["StockCode"] = df["StockCode"].astype("str")
df_not5digit = df[(df["StockCode"].str.match("^\\d{5}$") == False)]
df_not5digit.loc[:,"StockCode"].unique()

array(['79323P', '79323W', '48173C', ..., '85049g', '72803b', '79151B'],
      shape=(1675,), dtype=object)

In [35]:
# It appears that there are many StockCode that didn't follow the 5-digit patterns, but many StockCode ends with a Letter. Let's examine if there are other patterns
df_not5digit_endwletter = df[(df["StockCode"].str.match("^\\d{5}$") == False) & ((df["StockCode"].str.match("^\\d{5}[a-zA-Z]+$") == False))]
df_not5digit_endwletter.loc[:,"StockCode"].unique()

array(['POST', 'D', 'DCGS0058', 'DCGS0068', 'DOT', 'M', 'DCGS0004',
       'DCGS0076', 'C2', 'BANK CHARGES', 'DCGS0003', 'TEST001',
       'gift_0001_80', 'DCGS0072', 'gift_0001_20', 'DCGS0044', 'TEST002',
       'gift_0001_10', 'gift_0001_50', 'DCGS0066N', 'gift_0001_30',
       'PADS', 'ADJUST', 'gift_0001_40', 'gift_0001_60', 'gift_0001_70',
       'gift_0001_90', 'DCGSSGIRL', 'DCGS0006', 'DCGS0016', 'DCGS0027',
       'DCGS0036', 'DCGS0039', 'DCGS0060', 'DCGS0056', 'DCGS0059', 'GIFT',
       'DCGSLBOY', 'm', 'DCGS0053', 'DCGS0062', 'DCGS0037', 'DCGSSBOY',
       'DCGSLGIRL', 'S', 'DCGS0069', 'DCGS0070', 'DCGS0075', 'B',
       'DCGS0041', 'ADJUST2', '47503J ', 'C3', 'SP1002', 'AMAZONFEE'],
      dtype=object)

In [36]:
# Let's examine each unusual StockCode to understand what each represent
df[df["StockCode"].str.contains("POST")] # I will change the value in the "contains" for each unique unusual StockCode and summarize below

Invoice StockCode Description  Quantity         InvoiceDate  \
89       489439      POST     POSTAGE         3 2009-12-01 09:28:00   
126      489444      POST     POSTAGE         1 2009-12-01 09:55:00   
173      489447      POST     POSTAGE         1 2009-12-01 10:10:00   
625      489526      POST     POSTAGE         6 2009-12-01 11:50:00   
927     C489538      POST     POSTAGE        -1 2009-12-01 12:18:00   
...         ...       ...         ...       ...                 ...   
522916   537967      POST     POSTAGE         1 2010-12-09 11:31:00   
522930   537995      POST     POSTAGE         2 2010-12-09 11:39:00   
522974   538003      POST     POSTAGE         8 2010-12-09 12:05:00   
523064   538008      POST     POSTAGE         5 2010-12-09 12:17:00   
524491   538093      POST     POSTAGE         5 2010-12-09 14:49:00   

                      Price          Customer ID         Country  
89                    18.00             12682.00          France  
126                  141.00             12636.00             USA  
173                  130.00             12362.00         Belgium  
625                   18.00             12533.00         Germany  
927                    9.58             15796.00  United Kingdom  
...                     ...                  ...             ...  
522916                18.00             12494.00          France  
522930                18.00             12712.00         Germany  
522974                18.00             12429.00         Denmark  
523064                18.00             12683.00          France  
524491                18.00             12682.00          France  

[865 rows x 8 columns]

In [37]:
df[df["StockCode"].str.contains("^DOT$")]

Invoice StockCode     Description  Quantity         InvoiceDate  \
2379    489597       DOT  DOTCOM POSTAGE         1 2009-12-01 14:28:00   
2539    489600       DOT  DOTCOM POSTAGE         1 2009-12-01 14:43:00   
2551    489601       DOT  DOTCOM POSTAGE         1 2009-12-01 14:44:00   
2571    489602       DOT  DOTCOM POSTAGE         1 2009-12-01 14:45:00   
2619    489603       DOT  DOTCOM POSTAGE         1 2009-12-01 14:46:00   
...        ...       ...             ...       ...                 ...   
524272  538071       DOT  DOTCOM POSTAGE         1 2010-12-09 14:09:00   
524887  538148       DOT  DOTCOM POSTAGE         1 2010-12-09 16:26:00   
525000  538149       DOT  DOTCOM POSTAGE         1 2010-12-09 16:27:00   
525126  538153       DOT  DOTCOM POSTAGE         1 2010-12-09 16:31:00   
525147  538154       DOT  DOTCOM POSTAGE         1 2010-12-09 16:35:00   

                      Price  Customer ID         Country  
2379                 647.19          NaN  United Kingdom  
2539                  55.96          NaN  United Kingdom  
2551                  68.39          NaN  United Kingdom  
2571                  59.35          NaN  United Kingdom  
2619                  42.39          NaN  United Kingdom  
...                     ...          ...             ...  
524272               885.94          NaN  United Kingdom  
524887               547.32          NaN  United Kingdom  
525000               620.68          NaN  United Kingdom  
525126               822.94          NaN  United Kingdom  
525147                85.79          NaN  United Kingdom  

[736 rows x 8 columns]

In [38]:
df[df["StockCode"].str.contains("^M$")]

Invoice StockCode Description  Quantity         InvoiceDate  \
2697     489609         M      Manual         1 2009-12-01 14:50:00   
3053    C489651         M      Manual        -1 2009-12-01 16:48:00   
5897    C489859         M      Manual        -1 2009-12-02 14:45:00   
9259    C490126         M      Manual        -1 2009-12-03 18:12:00   
9307    C490129         M      Manual        -1 2009-12-03 18:26:00   
...         ...       ...         ...       ...                 ...   
521074   537800         M      Manual        12 2010-12-08 13:08:00   
521122   537804         M      Manual        12 2010-12-08 13:17:00   
521124   537804         M      Manual        12 2010-12-08 13:17:00   
524696  C538124         M      Manual        -4 2010-12-09 15:43:00   
524776   538147         M      Manual         1 2010-12-09 16:11:00   

                      Price          Customer ID         Country  
2697                   4.00                  NaN  United Kingdom  
3053                   5.10             17804.00  United Kingdom  
5897                  69.57                  NaN  United Kingdom  
9259                   5.95             15884.00  United Kingdom  
9307                1998.49             15482.00  United Kingdom  
...                     ...                  ...             ...  
521074                 0.19             12748.00  United Kingdom  
521122                 2.95             12748.00  United Kingdom  
521124                 0.19             12748.00  United Kingdom  
524696                 0.50             15329.00  United Kingdom  
524776                15.00             13090.00  United Kingdom  

[850 rows x 8 columns]

## StockCode Unusual Codes

##### <li> StockCode is meant to follow the paterrn <strong> [0-9]{5} </strong> but seems to have legit values for <strong> [0-9]{5}[a-zA-Z]+</strong>. It also contains other values as listed below: </li>
| Code | Description | Impact if Included | Action |
| --- | --- | --- | --- |
| ADJUSTX | Looks like manual account adjustments by admins | Not tied to customer purchasing behavior| Exclude |
| AMAZONFEE | Looks like fees for Amazon shipping or something | Not tied to customer purchasing behavior| Exclude | 
| BANK CHARGES or B | Bank charges | Not tied to customer purchasing behavior| Exclude | 
| C2 | Carriage transaction - not sure what this means | Adds noise, not enough data to generalize | Exclude | 
| C3 | Not sure, only 1 transaction | Adds noise, not enough data to generalize |Exclude |
| DCGS | Looks valid, some quantities are negative though and customer ID is null | Can't link to customer profiles | Exclude | 
| D | Looks valid, represents discount values | Can't link to customer profiles | Exclude | 
| DOT | Looks valid, represents postage charges | Can't link to customer profiles| Exclude | 
| gift__XXX | Purchases with gift cards, might be interesting for another analysis, but no customer data | Can't link to customer profiles | Exclude |
| M or m | Looks valid, represents manual transactions | Can't link to customer profiles|  Exclude | 
| PADS | Looks like a legit stock code for padding | Tied to customer purchasing behavior | Include | 
| S | Samples sent to customer | Add noises, Not reflective of actual purchases | Exclude | 
| SP1002 | Looks like a special request item, only 2 transactions, 3 look legit, 1 has 0 pricing | Adds noise, not enough data to generalize | Exclude | 
| TESTXXX | Testing data, not valid | Adds noise, not enough data to generalize | Exclude |

##### Why This Matters?
<li> Outliers and irrelevant data can skew centroids in Customer Segmentation and produce misleading clusters, thus removing them improve the quality and interpretability of customer clusters </li>


# Data Cleaning

In [39]:
# Create a new dataframe as a copy of our original dataframe
cleaned_df = df.copy()

In [40]:
# Clean the first column, Invoice

# Adjust data type to string so that we can apply regex as there are specific patterns that we want to filter out, aka those with letter "C" or "A"
cleaned_df["Invoice"] = cleaned_df["Invoice"].astype("str")

# Create a new variable "mask" as a filter expression
mask = (
    cleaned_df["Invoice"].str.match("^\\d{6}$") == True
)

# Apply filter "mask" to get Invoice with only 6-digit code -- here we got 515,252 records
cleaned_df = cleaned_df[mask]
cleaned_df

Invoice StockCode                          Description  Quantity  \
0       489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1       489434    79323P                   PINK CHERRY LIGHTS        12   
2       489434    79323W                  WHITE CHERRY LIGHTS        12   
3       489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4       489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
...        ...       ...                                  ...       ...   
525456  538171     22271                 FELTCRAFT DOLL ROSIE         2   
525457  538171     22750         FELTCRAFT PRINCESS LOLA DOLL         1   
525458  538171     22751       FELTCRAFT PRINCESS OLIVIA DOLL         1   
525459  538171     20970   PINK FLORAL FELTCRAFT SHOULDER BAG         2   
525460  538171     21931               JUMBO STORAGE BAG SUKI         2   

               InvoiceDate                Price          Customer ID  \
0      2009-12-01 07:45:00                 6.95             13085.00   
1      2009-12-01 07:45:00                 6.75             13085.00   
2      2009-12-01 07:45:00                 6.75             13085.00   
3      2009-12-01 07:45:00                 2.10             13085.00   
4      2009-12-01 07:45:00                 1.25             13085.00   
...                    ...                  ...                  ...   
525456 2010-12-09 20:01:00                 2.95             17530.00   
525457 2010-12-09 20:01:00                 3.75             17530.00   
525458 2010-12-09 20:01:00                 3.75             17530.00   
525459 2010-12-09 20:01:00                 3.75             17530.00   
525460 2010-12-09 20:01:00                 1.95             17530.00   

               Country  
0       United Kingdom  
1       United Kingdom  
2       United Kingdom  
3       United Kingdom  
4       United Kingdom  
...                ...  
525456  United Kingdom  
525457  United Kingdom  
525458  United Kingdom  
525459  United Kingdom  
525460  United Kingdom  

[515252 rows x 8 columns]

In [42]:
# Adjust a variable "mask" as a filter expression for StockCode column
cleaned_df["StockCode"] = cleaned_df["StockCode"].astype("str")

mask = (
    (cleaned_df["StockCode"].str.match("^\\d{5}$") == True)
    | (cleaned_df["StockCode"].str.match("^\\d{5}[a-zA-Z]+$") == True) # Matches StockCode that starts with one-or-more upper-or lower case letters and also ends with it
    | (cleaned_df["StockCode"].str.match("^PADS$") == True)
)   

cleaned_df = cleaned_df[mask]
cleaned_df

C:\Users\JohnP\AppData\Local\Temp\ipykernel_23036\3693338055.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cleaned_df["StockCode"] = cleaned_df["StockCode"].astype("str")


Invoice StockCode                          Description  Quantity  \
0       489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1       489434    79323P                   PINK CHERRY LIGHTS        12   
2       489434    79323W                  WHITE CHERRY LIGHTS        12   
3       489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4       489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
...        ...       ...                                  ...       ...   
525456  538171     22271                 FELTCRAFT DOLL ROSIE         2   
525457  538171     22750         FELTCRAFT PRINCESS LOLA DOLL         1   
525458  538171     22751       FELTCRAFT PRINCESS OLIVIA DOLL         1   
525459  538171     20970   PINK FLORAL FELTCRAFT SHOULDER BAG         2   
525460  538171     21931               JUMBO STORAGE BAG SUKI         2   

               InvoiceDate                Price          Customer ID  \
0      2009-12-01 07:45:00                 6.95             13085.00   
1      2009-12-01 07:45:00                 6.75             13085.00   
2      2009-12-01 07:45:00                 6.75             13085.00   
3      2009-12-01 07:45:00                 2.10             13085.00   
4      2009-12-01 07:45:00                 1.25             13085.00   
...                    ...                  ...                  ...   
525456 2010-12-09 20:01:00                 2.95             17530.00   
525457 2010-12-09 20:01:00                 3.75             17530.00   
525458 2010-12-09 20:01:00                 3.75             17530.00   
525459 2010-12-09 20:01:00                 3.75             17530.00   
525460 2010-12-09 20:01:00                 1.95             17530.00   

               Country  
0       United Kingdom  
1       United Kingdom  
2       United Kingdom  
3       United Kingdom  
4       United Kingdom  
...                ...  
525456  United Kingdom  
525457  United Kingdom  
525458  United Kingdom  
525459  United Kingdom  
525460  United Kingdom  

[512796 rows x 8 columns]

In [43]:
# "Description" Column shouldn't have any issue
# "Quantity" column may contain negative values, and from our EDA, negative values tend to belong to Null "CustomerID".

# Remove Null CustomerID
cleaned_df.dropna(subset=["Customer ID"], inplace = True)
cleaned_df.describe()

Quantity                    InvoiceDate  \
count            406337.00                         406337   
mean                 13.62  2010-07-01 10:11:06.543288320   
min                   1.00            2009-12-01 07:45:00   
25%                   2.00            2010-03-26 14:01:00   
50%                   5.00            2010-07-09 15:48:00   
75%                  12.00            2010-10-14 17:09:00   
max               19152.00            2010-12-09 20:01:00   
std                  97.00                            NaN   

                     Price          Customer ID  
count            406337.00            406337.00  
mean                  2.99             15373.63  
min                   0.00             12346.00  
25%                   1.25             14004.00  
50%                   1.95             15326.00  
75%                   3.75             16814.00  
max                 295.00             18287.00  
std                   4.29              1677.37

In [44]:
# "Price" Column appears to have 0 as minimum, which may impact our customer analysis. Let's check how many instances of 0 price and remove them
len(cleaned_df[cleaned_df["Price"] == 0])

28

In [45]:
# Only 28 instances of 0 price so it is okay to drop these records
cleaned_df = cleaned_df[cleaned_df["Price"] > 0]

cleaned_df.describe()

Quantity                    InvoiceDate  \
count            406309.00                         406309   
mean                 13.62  2010-07-01 10:14:25.869572352   
min                   1.00            2009-12-01 07:45:00   
25%                   2.00            2010-03-26 14:01:00   
50%                   5.00            2010-07-09 15:48:00   
75%                  12.00            2010-10-14 17:09:00   
max               19152.00            2010-12-09 20:01:00   
std                  97.00                            NaN   

                     Price          Customer ID  
count            406309.00            406309.00  
mean                  2.99             15373.72  
min                   0.00             12346.00  
25%                   1.25             14006.00  
50%                   1.95             15326.00  
75%                   3.75             16814.00  
max                 295.00             18287.00  
std                   4.29              1677.33

In [46]:
# It appears that there are instances whose prices are around 0. Let's check them
print(cleaned_df["Price"].min())
print(len(cleaned_df[cleaned_df["Price"] == cleaned_df["Price"].min()]))
# Since these $0.01 records account for only 14 rows, it is okay to include them

0.001
14


In [47]:
# Export cleaned dataframe for further customer analysis
cleaned_df.to_excel('cleaned_df.xlsx', index=False)

In [48]:
# Data cleaning is completed. Let's check how much data we have left
print(f"Data dropped after cleaning: {1 - len(cleaned_df)/len(df):.2%}")
print(f"Data remain after cleaning: {len(cleaned_df)/len(df):.2%}")
print(f"Rows dropped after cleaning: {df.shape[0] - cleaned_df.shape[0]:,} rows ({1 - len(cleaned_df)/len(df):.2%})")
print(f"Rows remain after cleaning: {cleaned_df.shape[0]:,} rows ({len(cleaned_df)/len(df):.2%})")

Data dropped after cleaning: 22.68%
Data remain after cleaning: 77.32%
Rows dropped after cleaning: 119,152 rows (22.68%)
Rows remain after cleaning: 406,309 rows (77.32%)


## Assumptions
Negative Quantity = Invalid Transaction
<li> Assumes negative quantities are errors or irrelevant (e.g., returns or adjustments). </li>
<li> Assumes they shouldn't be included in sales or customer analysis. </li>

Unusual StockCode = Non-Standard Product
<li>Assumes only codes following a strict pattern (e.g., 5-digit or 5-digit + letter, PADS) are valid.</li>
<li>Excludes promotional items, packaging, or services that may use different codes.</li>

Invoices with "C" or "A" = Non-Sales Events
<li>Assumes cancellations ("C") and accounting adjustments ("A") should be excluded from customer analysis.</li>

Null CustomerID = Non-Actionable Record
<li>Assumes records without customer IDs can't be linked to meaningful behavior or segmentation.</li>

## Questions to Ask Stakeholders

After applying initial cleaning rules: removing records with negative quantities, null customer IDs, unusual stock codes, and cancellation/accounting invoices, we’ve excluded about 23% of the data.
1. Are these exclusions acceptable for our analysis goals? 
2. Are there any edge cases or business rules we should reconsider before proceeding?

About Product Codes
1. Do all valid products follow the 5-digit or 5-digit + letter or PADS StockCode pattern?
2. Are there special codes in StockCode that we should retain?